File: coolrmi/hu.qgears.coolrmi/src/hu/qgears/coolrmi/serializer/ListSerializer.java
package hu.qgears.coolrmi.serializer;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

public class ListSerializer extends TypeSerializer {

	public ListSerializer() {
		super(TypeId.List, null, null);
	}

	@Override
	public boolean canSerializeIsSpecial() {
		return true;
	}

	@Override
	public boolean canSerialize(PortableSerializer serializer, JavaType typ) {
		return List.class.isAssignableFrom(typ.getCls());
	}

	public JavaType getElementType(JavaType typ) {
		return typ.getGenericTypes()[0];
	}

	@Override
	public void writeType(PortableSerializer serializer, OutputStream os,
			JavaType typ) throws IOException {
		super.writeType(serializer, os, typ);
		serializer.writeClassName(os, getElementType(typ));
	}

	@Override
	public JavaType readType(PortableSerializer serializer, InputStream is)
			throws IOException, ClassNotFoundException {
		JavaType items = serializer.readClassName(is);
		return new JavaType(ArrayList.class, new JavaType[] { items });
	}

	@Override
	public void serialize(PortableSerializer serializer, OutputStream os,
			Object o, JavaType typ) throws IOException {
		if (o == null) {
			Utils.write32(os, -1);
			return;
		}
		List<?> lst = (List<?>) o;
		Utils.write32(os, lst.size());

		JavaType elemType = getElementType(typ);
		for (Object el : lst) {
			serializer.serialize(os, el, elemType);
		}
	}

	@Override
	public Object deserialize(PortableSerializer serializer, InputStream is,
			JavaType typ) throws Exception {
		int len = Utils.read32(is);
		if (len == -1) {
			return null;
		}

		List<Object> lst = new ArrayList<Object>(len);
		JavaType elemType = getElementType(typ);

		for (int i = 0; i < len; ++i) {
			lst.add(serializer.deserialize(is, elemType));
		}
		return lst;
	}

}

File: coolrmi/CoolRMI.Net/Serializer/ListSerializer.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;

namespace CoolRMI.Net.Serializer
{
    public class ListSerializer : TypeSerializer
    {
         public ListSerializer() : base(TypeId.List, null) {}

        public override bool CanSerializeIsSpecial => true;
        public override bool CanSerialize(PortableSerializer serializer, Type typ)
        {
            return typeof(IList<>).IsAssignableFrom(typ);
        }

        private static Type GetElementType(Type typ)
        {
            return typ.GetInterface("System.Collections.Generic.IList`1")
                .GetGenericArguments()[0];
        }

        public override void WriteType(PortableSerializer serializer,
            BinaryWriter bw, Type typ)
        {
            base.WriteType(serializer, bw, typ);
            serializer.WriteClassName(bw, GetElementType(typ));
        }

        public override Type ReadType(PortableSerializer serializer, BinaryReader br)
        {
            var items = serializer.ReadClassName(br);
            return typeof(List<>).MakeGenericType(items);
        }

        public override void Serialize(PortableSerializer serializer,
            BinaryWriter bw, object o, Type typ)
        {
            if (o == null)
            {
                bw.Write(-1);
                return;
            }

            // todo: IList<> csak IEnumerable-t implemental, nem IList-et...
            var lst = (IList) o;
            bw.Write(lst.Count);

            var elemType = GetElementType(typ);
            foreach (var el in lst)
                serializer.Serialize(bw, el, elemType);
        }

        public override object Deserialize(PortableSerializer serializer,
            BinaryReader br, Type typ)
        {
            var len = br.ReadInt32();
            if (len == -1) return null;

            var inst = (IList) Activator.CreateInstance(typ, len);
            var elemType = GetElementType(typ);

            for (int i = 0; i < len; ++i)
                inst.Add(serializer.Deserialize(br, elemType));

            return inst;
        }
    }
}
